Source/HTTP/WinHttp/winhttp_http_task.cpp

```diff
@@ -30,29 +30,165 @@ void get_proxy_name(
+        sizeof(WINHTTP_PROXY_INFO));
+    if (!result)
+    {
+        DWORD dwError = GetLastError();
+        HC_TRACE_ERROR(HTTPCLIENT, "Internal_SetGlobalProxy [TID %ul] WinHttpSetOption errorcode %d", GetCurrentThreadId(), dwError);
+        return E_FAIL;
+    }
+
+    return S_OK;
+}
+
+XTaskQueueHandle HC_PERFORM_ENV::GetImmediateQueue()
+{
+    std::lock_guard<std::mutex> lock(m_lock);
+    if (m_immediateQueue == nullptr)
+    {
+        (void)XTaskQueueCreate(XTaskQueueDispatchMode::Immediate, XTaskQueueDispatchMode::Immediate, &m_immediateQueue);
+    }
+
+    return m_immediateQueue;
+}
+
+uint32_t HC_PERFORM_ENV::GetDefaultHttpSecurityProtocolFlagsForWin7()
+{
+    uint32_t enabledHttpSecurityProtocolFlags = WINHTTP_FLAG_SECURE_PROTOCOL_TLS1 | WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_1 | WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_2;
+    return enabledHttpSecurityProtocolFlags;
+}
+
+HINTERNET HC_PERFORM_ENV::CreateHSessionForForHttpSecurityProtocolFlags(_In_ uint32_t enabledHttpSecurityProtocolFlags)
+{
+    xbox::httpclient::Uri proxyUri;
+    DWORD accessType{ WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY };
     http_internal_wstring wProxyName;
+    m_proxyType = get_ie_proxy_info(proxy_protocol::https, proxyUri);
     get_proxy_name(m_proxyType, proxyUri, &accessType, &wProxyName);
 
-    m_hSession = WinHttpOpen(
+    HINTERNET hSession = WinHttpOpen(
         nullptr,
         accessType,
         wProxyName.length() > 0 ? wProxyName.c_str() : WINHTTP_NO_PROXY_NAME,
         WINHTTP_NO_PROXY_BYPASS,
-        WINHTTP_FLAG_ASYNC);
+#if HC_PLATFORM == HC_PLATFORM_GDK
+        WINHTTP_FLAG_SECURE_DEFAULTS
+#else
+        WINHTTP_FLAG_ASYNC
+#endif
+    );
+
+#if HC_PLATFORM == HC_PLATFORM_GDK
+    DWORD error = GetLastError();
+    if (error == ERROR_INVALID_PARAMETER)
+    {
+        // This might happen on older Win10 PC versions that don't support WINHTTP_FLAG_SECURE_DEFAULTS
+        hSession = WinHttpOpen(
+            nullptr,
+            WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY,
+            WINHTTP_NO_PROXY_NAME,
+            WINHTTP_NO_PROXY_BYPASS,
+            WINHTTP_FLAG_ASYNC);
+    }
+#endif
+
+    if (hSession == nullptr)
+    {
+        DWORD dwError = GetLastError();
+        HC_TRACE_ERROR(HTTPCLIENT, "HC_PERFORM_ENV WinHttpOpen errorcode %d", dwError);
+        return nullptr;
+    }
+
+    auto result = WinHttpSetOption(
+        hSession,
+        WINHTTP_OPTION_SECURE_PROTOCOLS,
+        &enabledHttpSecurityProtocolFlags,
+        sizeof(enabledHttpSecurityProtocolFlags));
+    if (!result)
+    {
+        DWORD dwError = GetLastError();
+        HC_TRACE_ERROR(HTTPCLIENT, "HC_PERFORM_ENV WinHttpSetOption errorcode %d", dwError);
+        return nullptr;
+    }
+
+    if (!m_globalProxy.empty())
+    {
+        (void)SetGlobalProxyForHSession(hSession, m_globalProxy.c_str());
+    }
+
+    return hSession;
+}
+
+HINTERNET HC_PERFORM_ENV::GetSessionForHttpSecurityProtocolFlags(_In_ uint32_t enabledHttpSecurityProtocolFlags)
```

Comment:
>GetSessionForHttpSecurityProtocolFlags [](start = 26, length = 38)

Any reason to split this function and CreateSession above? looks like you always want to create the session if it doesn't exist yet

Reply:
It was just to make it easier to read GetSessionForHttpSecurityProtocolFlags as this function just dealing with the std::map